addons/product/models/product_template.py

```diff
@@ -938,6 +939,83 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, product_template_attribute_values, parent_combination):
+        """
+        Generate all possible combination for attributes values (aka cartesian product).
+        It is equivalent to itertools.product except it skips invalid partial combinations before they are complete.
+
+        Imagine the cartesian product of 'A', 'CD' and range(1_000_000) and let's say that 'A' and 'C' are incompatible.
+        If you use itertools.product or any normal cartesian product, you'll need to filter out of the final result
+        the 1_000_000 combinations that start with 'A' and 'C' . Instead, This implementation will test if 'A' and 'C' are
+        compatible before even considering range(1_000_000), skip it and and continue with combinations that start
+        with 'A' and 'D'.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+
+        :param product_template_attribute_values: the values we want all the possibles combinations of.
+        One list of values by attribute line
+        :return: a generator of product template values combinations (in the form of lists)
+        """
+        if not product_template_attribute_values:
+            return
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # The following dict uses product template attribute values as keys
+        # 0 means the value is acceptable, greater than 0 means it's rejected, it cannot be negative
+        # Bear in mind that several values can reject the same value and the latter can only be included in the
+        #  considered combination if no value rejects it.
+        # This dictionary counts how many times each value is rejected.
+        # Each time a value is included in the considered combination, the values it rejects are incremented
+        # When a value is discarded from the considered combination, the values it rejects are decremented
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values)
+        # determines which attribute line we're working on
+        attribute_idx = 0
+
+        def current_value_id():
+            return product_template_attribute_values[attribute_idx][values_idx[attribute_idx]].id
```

Comment:
Ideally we would not need this helper. What would be best is if the "current value" could be assigned once at the start of the loop in a variable, and every time it changes, the loop would be restarted.  Because it's hard to follow when the iterating variable changes in the middle of the loop.

I think this is more or less already the case, hopefully it's possible to make few adaptations to reach this goal, and then the whole process would probably be much more clear.

Reply:
There are two ways to iterate in this loop : 
1. Select the next value for the current attribute
2. Consider the next attribute
```
# first we try to get to next value of current attribute
[...]
# then we get to the next attribute or submit combination if we're on the last attribute
```

It's not possible to assign a single value at the start of the loop.

Reply:
I see what you mean but there must be a way to make it flow more logically.

It's fine if the value is assigned after the conditions, so long as it's just once, and the value is not accessed before that point.

It's strange to do the discard at the start of the loop, instead of just after a value has been selected.
When first reading the code it would be much more natural to read it like this:
- select next value (and assign it once for the rest of the loop)
- apply exclusions based on the selected value
- check if result is valid

Reply:
I've removed the helper function and replaced it with two assignments.

In the following UML diagram, the red arrows shows where `current_ptav` is modified and the blue arrows show where it's assigned. I don't think trying to get to one assignment will get anything clearer.
![rejection-cartesian](https://user-images.githubusercontent.com/7291317/106917305-6c576c00-6708-11eb-873c-fc685f68ccca.png)